src/circuit/layouter.rs

```diff
@@ -102,26 +104,122 @@ impl<'a, C: Chip, CS: Assignment<C::Field> + 'a> Layouter<C> for SingleChip<'a,
 
     fn assign_region(
         &mut self,
-        assignment: impl FnOnce(Region<'_, C>) -> Result<(), Error>,
+        mut assignment: impl FnMut(Region<'_, C>) -> Result<(), Error>,
     ) -> Result<(), Error> {
         let region_index = self.regions.len();
-        self.regions.push(self.current_gate);
+
+        // Get shape of the region.
+        let mut shape = RegionShape::new(region_index);
+        {
+            let region: &mut dyn RegionLayouter<C> = &mut shape;
+            assignment(region.into())?;
+        }
+
+        // Lay out this region. We implement the simplest approach here: position the
+        // region starting at the earliest row for which none of the columns are in use.
+        let mut region_start = 0;
+        for column in &shape.columns {
+            region_start = cmp::max(region_start, self.columns.get(column).cloned().unwrap_or(0));
+        }
+        self.regions.push(region_start);
+
+        // Update column usage information.
+        for column in shape.columns {
+            self.columns.insert(column, region_start + shape.row_count);
+        }
 
         let mut region = SingleChipRegion::new(self, region_index);
         {
             let region: &mut dyn RegionLayouter<C> = &mut region;
             assignment(region.into())?;
         }
-        self.current_gate += region.row_count;
 
         Ok(())
     }
 }
 
+/// The shape of a region. For a region at a certain index, we track
+/// the set of columns it uses as well as the number of rows it uses.
+#[derive(Debug)]
+pub struct RegionShape {
+    region_index: usize,
```

Comment:
The concept I had of a region shape (e.g. for #93) didn't include the starting index, but I guess that can be changed later if needed.

Reply:
This isn't the row that the region starts on; it's a "region ID" within the layouter, used to enable `Cell`s to be pointers to regions instead of to absolute circuit positions (which directly enables the kind of layouting that this PR implements). `RegionShape` stores `region_index` solely to ensure that any `Cell`s it returns are consistent across both the shape-finding pass, and the assignment pass.